src/lucky/param_serializable.cr

```diff
@@ -0,0 +1,177 @@
+                ERROR
+              else
+                @{{ metadata[:type_declaration].var.id }} = Lucky::PermittedParam({{ metadata[:type_declaration].type }}).new(name: {{ metadata[:type_declaration].var.stringify }}, value: default_or_nil, param_key: %param_key_value)
+              end
+            {% end %}
+          else
+            {% if metadata[:is_serializable] %}
+            @{{ metadata[:type_declaration].var.id }} = Lucky::PermittedParam({{ metadata[:type_declaration].type }}).new(name: {{ metadata[:type_declaration].var.stringify }}, value: {{ metadata[:core_type] }}.from_params(%val.as(Lucky::Params)), param_key: %param_key_value)
+            {% else %}
+              # NOTE: these come from Avram directly
+              %result = {{ metadata[:core_type] }}::Lucky.parse(%val)
+
+              if %result.is_a? Avram::Type::SuccessfulCast
+                @{{ metadata[:type_declaration].var.id }} = Lucky::PermittedParam({{ metadata[:type_declaration].type }}).new(name: {{ metadata[:type_declaration].var.stringify }}, value: %result.value.as({{ metadata[:type] }}), param_key: %param_key_value)
+              else
+                raise Lucky::InvalidParamError.new(
+                  param_name: "{{ metadata[:type_declaration].var.id }}",
+                  param_value: %val.to_s,
+                  param_type: "{{ metadata[:type] }}"
+                )
+              end
+            {% end %}
+          end
+        {% end %}
+      {% end %}
+    end
+
+    macro param(type_declaration, param_key = nil)
+      {% unless type_declaration.is_a?(TypeDeclaration) %}
+        {% raise "'param' expects a type declaration like 'name : String', instead got: '#{type_declaration}'" %}
+      {% end %}
+
+      {% is_nilable_type = type_declaration.type.resolve.nilable? %}
+      {% type = is_nilable_type ? type_declaration.type.types.reject(&.==(Nil)).first.resolve : type_declaration.type.resolve %}
+      {% is_array = type.name.starts_with?("Array") %}
+      # TODO: This probably breaks if the type is Array(String?)
+      {% core_type = is_array ? type.type_vars.first : type %}
+      {% is_file = core_type.name.starts_with?("Lucky::UploadedFile") %}
+      {% is_param_serializable = core_type.resolve < Lucky::ParamSerializable %}
+
+      {%
+        PARAM_DECLARATIONS << {
+          type_declaration: type_declaration,
+          param_key:        param_key,
+          nilable:          is_nilable_type,
+          type:             type,
+          core_type:        core_type,
+          is_array:         is_array,
+          is_file:          is_file,
+          is_serializable:  is_param_serializable,
+        }
+      %}
+
+      getter {{ type_declaration.var }} : Lucky::PermittedParam({{ type_declaration.type }}){% if is_nilable_type %}?{% end %}
+
+      # TODO: This seems to always raise, even though I'm never calling it
+      # def {{ type_declaration.var }}=(val : {{ type_declaration.type }})
```

Comment:
This would be great to have so you don't try to do something like:

```crystal
user_params = UserParams.new
user_params.name = "Jeremy"
```

However, I couldn't get it to stop raising the exception... Maybe I missed something?